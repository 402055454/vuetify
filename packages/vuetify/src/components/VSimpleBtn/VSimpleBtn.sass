// Imports
@import './_variables.scss'

// Theme
.v-simple-btn:not(.v-simple-btn--outlined)
  &.primary,
  &.secondary,
  &.accent,
  &.success,
  &.error,
  &.warning,
  &.info
    color: map-deep-get($material-dark, 'text', 'primary')

+theme(v-simple-btn) using ($material)
  color: map-deep-get($material, 'text', 'primary')

  &.v-simple-btn--disabled
    color: map-deep-get($material, 'buttons', 'disabled') !important

    &:not(.v-simple-btn--flat):not(.v-simple-btn--text):not(.v-simple-btn--outlined)
      background-color: map-deep-get($material, 'buttons', 'focused') !important

  &:not(.v-simple-btn--flat):not(.v-simple-btn--text):not(.v-simple-btn--outlined)
    background-color: map-get($material, 'app-bar')

  &.v-simple-btn--outlined.v-simple-btn--text
    border-color: map-get($material, 'dividers')

  +states($material)

// Block
.v-simple-btn
  align-items: center
  border-radius: $simple-btn-border-radius
  display: inline-flex
  flex: 0 0 auto
  font-weight: $simple-btn-font-weight
  letter-spacing: $simple-btn-letter-spacing
  justify-content: center
  outline: 0
  position: relative
  text-decoration: none
  text-indent: $simple-btn-letter-spacing
  text-transform: $simple-btn-text-transform
  transition-duration: $simple-btn-transition-duration
  transition-property: box-shadow, transform, opacity
  transition-timing-function: $simple-btn-transition-fn
  user-select: none
  vertical-align: middle
  white-space: nowrap

  @each $name, $size in $simple-btn-font-sizes
    &.v-size--#{$name}
      font-size: $size

  &:before
    border-radius: inherit
    bottom: 0
    color: inherit
    content: ''
    left: 0
    opacity: 0
    pointer-events: none
    position: absolute
    right: 0
    top: 0
    transition: $simple-btn-transition

  &:before
    background-color: currentColor

  &:not(.v-simple-btn--disabled)
    will-change: box-shadow

  &:not(.v-simple-btn--round)
    @each $name, $size in $simple-btn-sizes
      &.v-size--#{$name}
        height: #{$size}px
        min-width: #{round($size * 1.777777777777778)}px // default ratio
        padding: 0 #{$size / 2.25}px

  > .v-btn__content
    color: inherit

.v-simple-btn__content
  align-items: center
  color: inherit
  display: flex
  flex: 1 0 auto
  justify-content: inherit
  line-height: normal
  position: relative

// Modifiers
.v-simple-btn:not(.v-simple-btn__text):not(.v-simple-btn--outlined)
  &.v-simple-btn--active:before
    opacity: $simple-btn-active-opacity

  &:hover:before
    opacity: $simple-btn-hover-opacity

  &:focus:before
    opacity: $simple-btn-focus-opacity

.v-simple-btn--absolute,
.v-simple-btn--fixed
  position: absolute

  &.v-simple-btn--right
    right: map-get($grid-gutters, 'lg')

  &.v-simple-btn--left
    left: map-get($grid-gutters, 'lg')

  &.v-simple-btn--top
    top: map-get($grid-gutters, 'lg')

  &.v-simple-btn--bottom
    bottom: map-get($grid-gutters, 'lg')

.v-simple-btn--block
  display: flex
  flex: 1 0 auto
  min-width: 100% !important
  max-width: auto

.v-simple-btn--contained
  +elevation(2)

  &:after
    +elevation(4)

  &:active
    +elevation(8)

.v-simple-btn--depressed
  box-shadow: none !important

.v-simple-btn--disabled
  box-shadow: none
  pointer-events: none

.v-simple-btn--fixed
  position: fixed

  .v-simple-btn__content
    opacity: 0

.v-simple-btn--outlined
    border: $simple-btn-outline-border-width solid currentColor

.v-simple-btn--outlined,
.v-simple-btn--flat,
.v-simple-btn--text
  background-color: transparent

.v-simple-btn--outlined,
.v-simple-btn--rounded
  &:before
    border-radius: inherit

.v-simple-btn--rounded
  border-radius: $simple-btn-rounded-border-radius

.v-simple-btn--tile
  border-radius: 0
