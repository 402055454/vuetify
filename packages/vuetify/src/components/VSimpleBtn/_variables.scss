@import '../../styles/styles.sass';
@import '../../styles/tools/_functions.sass';

$simple-btn-active-opacity: 0.18 !default;
$simple-btn-border-radius: $border-radius-root !default;
$simple-btn-focus-opacity: 0.24 !default;
$simple-btn-font-weight: 500 !default;
$simple-btn-hover-opacity: 0.08 !default;
$simple-btn-letter-spacing: .0892857143em !default;
$simple-btn-outline-border-width: thin !default;
$simple-btn-rounded-border-radius: 28px !default;
$simple-btn-text-transform: uppercase !default;
$simple-btn-transition-duration: 0.28s !default;
$simple-btn-transition-fn: map-get($transition, 'fast-out-slow-in') !default;
$simple-btn-transition: opacity 0.2s map-get($transition, 'ease-in-out') !default;

$simple-btn-sizes: () !default;
$simple-btn-sizes: map-deep-merge(
  (
    'x-small': 20,
    'small': 28,
    'default': 36,
    'large': 44,
    'x-large': 52
  ),
  $simple-btn-sizes
);

$simple-btn-font-sizes: () !default;
$simple-btn-font-sizes: map-deep-merge(
  (
    'x-small': .625rem,
    'small': .75rem,
    'default': .875rem,
    'large': .875rem,
    'x-large': 1rem
  ),
  $simple-btn-font-sizes
);
