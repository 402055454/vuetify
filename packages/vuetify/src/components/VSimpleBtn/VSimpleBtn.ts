// Styles
import './VSimpleBtn.sass'

// Extensions
import VSheet from '../VSheet'

// Components
import VProgressCircular from '../VProgressCircular'

// Mixins
import { factory as GroupableFactory } from '../../mixins/groupable'
import { factory as ToggleableFactory } from '../../mixins/toggleable'
import Positionable from '../../mixins/positionable'
import Sizeable from '../../mixins/sizeable'

// Utilities
import mixins, { ExtractVue } from '../../util/mixins'
import { breaking } from '../../util/console'

// Types
import { VNode, VNodeData } from 'vue'
import { PropType } from 'vue/types/options'
import Ripple, { RippleOptions } from '../../directives/ripple'

const baseMixins = mixins(
  VSheet,
  Positionable,
  Sizeable,
  GroupableFactory('btnToggle'),
  ToggleableFactory('inputValue')
  /* @vue/component */
)
interface options extends ExtractVue<typeof baseMixins> {
  $el: HTMLElement
}

export default baseMixins.extend<options>().extend({
  name: 'v-simple-btn',

  directives: {
    Ripple,
  },

  props: {
    block: Boolean,
    disabled: Boolean,
    depressed: Boolean,
    icon: Boolean,
    loading: Boolean,
    outlined: Boolean,
    retainFocusOnClick: Boolean,
    ripple: {
      type: [Boolean, Object],
      default: null,
    },
    rounded: Boolean,
    tag: {
      type: String,
      default: 'button',
    },
    text: Boolean,
    type: {
      type: String,
      default: 'button',
    },
    value: null as any as PropType<any>,
  },

  data: () => ({
    proxyClass: 'v-simple-btn--active',
  }),

  computed: {
    classes (): any {
      return {
        'v-btn': true,
        'v-btn--absolute': this.absolute,
        'v-btn--block': this.block,
        'v-btn--bottom': this.bottom,
        'v-btn--contained': this.contained,
        'v-btn--depressed': (this.depressed) || this.outlined,
        'v-btn--disabled': this.disabled,
        'v-btn--fixed': this.fixed,
        'v-btn--flat': this.isFlat,
        'v-btn--icon': this.icon,
        'v-btn--left': this.left,
        'v-btn--loading': this.loading,
        'v-btn--outlined': this.outlined,
        'v-btn--right': this.right,
        'v-btn--round': this.isRound,
        'v-btn--rounded': this.rounded,
        'v-btn--text': this.text,
        'v-btn--tile': this.tile,
        'v-btn--top': this.top,
        ...this.themeClasses,
        ...this.groupClasses,
        ...this.elevationClasses,
        ...this.sizeableClasses,
      }
    },
    contained (): boolean {
      return Boolean(
        !this.isFlat &&
        !this.depressed &&
        // Contained class only adds elevation
        // is not needed if user provides value
        !this.elevation
      )
    },
    computedRipple (): RippleOptions | boolean {
      const defaultRipple = this.icon ? { circle: true } : true
      if (this.disabled) return false
      else return this.ripple != null ? this.ripple : defaultRipple
    },
    isFlat (): boolean {
      return Boolean(
        this.icon ||
        this.text ||
        this.outlined
      )
    },
    isRound (): boolean {
      return Boolean(this.icon)
    },
    styles (): object {
      return {
        ...this.measurableStyles,
      }
    },
  },

  created () {
    const breakingProps = [
      ['flat', 'text'],
      ['outline', 'outlined'],
      ['round', 'rounded'],
    ]

    /* istanbul ignore next */
    breakingProps.forEach(([original, replacement]) => {
      if (this.$attrs.hasOwnProperty(original)) breaking(original, replacement, this)
    })
  },

  methods: {
    click (e: MouseEvent): void {
      !this.retainFocusOnClick && e.detail && this.$el.blur()
      this.$emit('click', e)

      this.btnToggle && this.toggle()
    },
    genContent (): VNode {
      return this.$createElement('span', {
        staticClass: 'v-btn__content',
      }, this.$slots.default)
    },
    genLoader (): VNode {
      return this.$createElement('span', {
        class: 'v-btn__loader',
      }, this.$slots.loader || [this.$createElement(VProgressCircular, {
        props: {
          indeterminate: true,
          size: 23,
          width: 2,
        },
      })])
    },
  },

  render (h): VNode {
    const children = [
      this.genContent(),
      this.loading && this.genLoader(),
    ]
    const setColor = !this.isFlat ? this.setBackgroundColor : this.setTextColor
    const data: VNodeData = {
      class: this.classes,
      style: this.styles,
      props: {},
      directives: [{
        name: 'ripple',
        value: this.computedRipple,
      }],
      on: {
        click: this.click,
      },
    }

    return h(this.tag, this.disabled ? data : setColor(this.color, data), children)
  },
})
