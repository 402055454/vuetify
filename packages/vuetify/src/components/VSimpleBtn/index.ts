import VSimpleBtn from './VSimpleBtn'

export { VSimpleBtn }
export default VSimpleBtn
