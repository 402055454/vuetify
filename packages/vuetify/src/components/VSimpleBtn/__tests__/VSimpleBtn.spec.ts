// Libraries
import Vue from 'vue'

// Components
import VSimpleBtn from '../VSimpleBtn'

// Utilities
import {
  createLocalVue,
  mount,
  Wrapper,
  createWrapper,
} from '@vue/test-utils'
import { compileToFunctions } from 'vue-template-compiler'

describe('VSimpleBtn', () => {
  let mountFunction: (options?: object) => Wrapper<Vue>
  let localVue: typeof Vue

  beforeEach(() => {
    localVue = createLocalVue()

    mountFunction = (options = {}) => {
      return mount(VSimpleBtn, {
        localVue,
        ...options,
      })
    }
  })

  it('should render component and match snapshot', () => {
    expect(mountFunction().html()).toMatchSnapshot()
  })

  it('should render component with a color prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        color: 'red',
      },
    }).html()).toMatchSnapshot()

    expect(mountFunction({
      propsData: {
        color: 'red lighten-1',
        text: true,
      },
    }).html()).toMatchSnapshot()
  })

  it('should render component with tile prop and match snapshot', () => {
    const wrapper = mountFunction({
      propsData: {
        tile: true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should register and unregister', () => {
    const register = jest.fn()
    const unregister = jest.fn()

    const wrapper = mountFunction({
      provide: {
        btnToggle: {
          register,
          unregister,
        },
      },
    })

    expect(register).toHaveBeenCalled()
    wrapper.destroy()
    expect(unregister).toHaveBeenCalled()
  })

  it('should emit a click event', () => {
    const wrapper = mountFunction()
    const click = jest.fn()

    wrapper.vm.$on('click', click)
    wrapper.trigger('click')

    expect(click).toHaveBeenCalled()
  })

  it('should have v-simple-btn--depressed class when using depressed prop', () => {
    const wrapper = mountFunction({
      propsData: {
        depressed: true,
      },
    })

    expect(wrapper.classes('v-simple-btn--depressed')).toBe(true)
  })

  it('should have v-simple-btn--flat class when using text and depressed props', () => {
    const wrapper = mountFunction({
      propsData: {
        depressed: true,
        text: true,
      },
    })

    expect(wrapper.classes('v-simple-btn--flat')).toBe(true)
  })

  it('should have v-simple-btn--outlined and v-simple-btn--depressed when using outlined prop', () => {
    const wrapper = mountFunction({
      propsData: {
        outlined: true,
      },
    })

    expect(wrapper.classes('v-simple-btn--outlined')).toBe(true)
    expect(wrapper.classes('v-simple-btn--depressed')).toBe(true)
  })

  it('should have the correct elevation', () => {
    const wrapper = mountFunction()

    wrapper.setProps({ disabled: true })
    expect(wrapper.classes('elevation-2')).toBe(false)
    expect(wrapper.classes('v-simple-btn--disabled')).toBe(true)

    wrapper.setProps({ disabled: false, elevation: 24 })
    expect(wrapper.classes('elevation-24')).toBe(true)

    wrapper.setProps({ elevation: 2 })
    expect(wrapper.classes('elevation-2')).toBe(true)
  })

  it('should not add color classes if disabled', () => {
    const wrapper = mountFunction({
      propsData: {
        color: 'red lighten-1',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({
      disabled: true,
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should stringify non string|number values', () => {
    const wrapper = mountFunction({
      propsData: {
        value: 'foo',
      },
    })

    expect(wrapper.attributes('value')).toBe('foo')

    wrapper.setProps({ value: 2 })
    expect(wrapper.attributes('value')).toBe('2')

    wrapper.setProps({ value: { foo: 'bar' } })
    expect(wrapper.attributes('value')).toBe('{"foo":"bar"}')
  })

  it('should call toggle when used in button group', () => {
    const register = jest.fn()
    const unregister = jest.fn()
    const toggle = jest.fn()
    const wrapper = mountFunction({
      provide: {
        btnToggle: { register, unregister },
      },
      methods: { toggle },
    })

    wrapper.trigger('click')
    expect(toggle).toHaveBeenCalled()
  })

  it('should retain focus when clicked', () => {
    const wrapper = mountFunction({
      propsData: {
        retainFocusOnClick: true,
      },
    })
    const event = new MouseEvent('click', { detail: 1 })
    const blur = jest.fn()

    wrapper.element.blur = blur
    wrapper.element.dispatchEvent(event)

    expect(blur).not.toHaveBeenCalled()

    wrapper.setProps({ retainFocusOnClick: false })
    wrapper.element.dispatchEvent(event)

    expect(blur).toHaveBeenCalled()
  })

  it('should not ripple when disabled', () => {
    const wrapper = mountFunction()

    expect((wrapper.vm as any).computedRipple).toBe(true)

    wrapper.setProps({ disabled: true })
    expect((wrapper.vm as any).computedRipple).toBe(false)

    wrapper.setProps({ disabled: false, ripple: 'circle' })
    expect((wrapper.vm as any).computedRipple).toBe('circle')
  })
})
