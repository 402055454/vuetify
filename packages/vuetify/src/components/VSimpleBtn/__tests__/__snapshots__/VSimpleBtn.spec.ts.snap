// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VSimpleBtn should not add color classes if disabled 1`] = `
<button class="v-simple-btn v-simple-btn--contained theme--light v-size--default red lighten-1">
  <span class="v-simple-btn__content">
  </span>
</button>
`;

exports[`VSimpleBtn should not add color classes if disabled 2`] = `
<button class="v-simple-btn v-simple-btn--contained v-simple-btn--disabled theme--light v-size--default"
        disabled="disabled"
>
  <span class="v-simple-btn__content">
  </span>
</button>
`;

exports[`VSimpleBtn should render component and match snapshot 1`] = `
<button class="v-simple-btn v-simple-btn--contained theme--light v-size--default">
  <span class="v-simple-btn__content">
  </span>
</button>
`;

exports[`VSimpleBtn should render component with a color prop and match snapshot 1`] = `
<button class="v-simple-btn v-simple-btn--contained theme--light v-size--default red">
  <span class="v-simple-btn__content">
  </span>
</button>
`;

exports[`VSimpleBtn should render component with a color prop and match snapshot 2`] = `
<button class="v-simple-btn v-simple-btn--flat v-simple-btn--text theme--light v-size--default red--text text--lighten-1">
  <span class="v-simple-btn__content">
  </span>
</button>
`;

exports[`VSimpleBtn should render component with tile prop and match snapshot 1`] = `
<button class="v-simple-btn v-simple-btn--contained v-simple-btn--tile theme--light v-size--default">
  <span class="v-simple-btn__content">
  </span>
</button>
`;
